### DISCUSSION TITLE
xarray_ Group by season

### DISCUSSION THREAD

**Discussion by Bjrj92** [pydata/xarray#6664] [Q&A] [answered]
In xarray seasonal grouping. I need to group by following group of months: JF,MAM,JJAS,OND and sum the grouped season values to perform trend analysis, whereas default groupby.season months are DJF,MAM,JJA,SON.

Code:
```python
import xarray as xr
import numpy as np
import pymannkendall as mk

Data = xr.open_dataset(in_file)
print(Data)
```

```
<xarray.Dataset>
Dimensions:  (time: 43830, lat: 129, lon: 135)
Coordinates:
  * lat      (lat) float64 6.5 6.75 7.0 7.25 7.5 ... 37.5 37.75 38.0 38.25 38.5
  * lon      (lon) float64 66.5 66.75 67.0 67.25 67.5 ... 99.25 99.5 99.75 100.0
  * time     (time) datetime64[ns] 1901-01-01 1901-01-02 ... 2020-12-31
Data variables:
    rain     (time, lat, lon) float64 ...
Attributes:
    Conventions:  CF-1.7
    title:        IMD gridded data
    source:       https://imdpune.gov.in/
    history:      2022-06-02 04:43:16.476268 Python
    references:   
    comment:      
    crs:          epsg:4326

```

```python
Data = Data['rain'].groupby('time.season').sum('time')
Data = Data.sel(lon= np.arange(75,81,0.25), lat = np.arange(8, 14,0.25),method ='nearest')

output=[]
for i in np.arange(len(Data.lat.values)):
    for j in np.arange(len(Data.lon.values)):
        
        try:
            slope_val = mk.original_test(Data[:,i,j]).z
        except:
            slope_val=-9999
        output.append(slope_val)
output = np.copy(output).reshape(Data.lat.size,Data.lon.size)
score  = xr.DataArray(output, dims=('lat','lon'), coords={'lat':Data.lat, 'lon':Data.lon}, name='ZScore')
df = score.to_dataframe()
```

**Comment by mathause** [comment]
You can pass a DataArray with your self-defined groups to [GroupBy](https://xarray.pydata.org/en/stable/generated/xarray.DataArray.groupby.html) . See #6180 on one way how to define the groups (no need to apply the `dot` trick).

**Reply by keewis** [reply]
see also #6012 for a very similar question

**Reply by Bjrj92** [reply]
Not Working

**Reply by mathause** [reply]
Ok

**Reply by Bjrj92** [reply]
Any Solution?

**Reply by mathause** [reply]
We do not know what does not work for you - so we cannot help. Please see also #5404.

**Reply by Bjrj92** [reply]
I followed #5134 and  tried to regroup the default season pattern DJF,MAM,JJA,SON into JF,MAM,JJAS,OND.

```python
import numpy as np
import xarray as xr
import pymannkendall as mk
Data = xr.tutorial.open_dataset("air_temperature")
months = Data.time.dt.month
seasons = xr.full_like(months, fill_value="none", dtype="U4")
seasons.name = "season"
seasons[months.isin([1, 2])] = "JF"
seasons[months.isin([3, 4, 5])] = "MAM"
seasons[months.isin([6, 7, 8,9])] = "JJAS"
seasons[months.isin([10, 11, 12])] = "OND"
x = Data.groupby(seasons).mean().reindex(season =['JF','MAM','JJAS','OND'])
x = x['air'].groupby('season').sum('season')
output =[]
for i in np.arange(len(x.lat.values)):
       for j in np.arange(len(x.lon.values)):
           
           try:
               slope_val = mk.original_test(x[:,i,j]).trend           
             
           except:
               slope_val=-9999
           output.append(slope_val)
output = np.copy(output).reshape(x.lat.size,x.lon.size)
score  = xr.DataArray(output, dims=('lat','lon'), coords={'lat':x.lat, 'lon':x.lon}, name='trend')
df1 = score.to_dataframe() 
```
The resultant dataframe is obtained for one season. I need the seasonal sum of each year to perform trend analysis and resultant dataframe for 4 seasons.

In the codeline12: I tried to sum for each season 
`x = x['rain'].groupby('season').sum('season.JF')`


ValueError: cannot reduce over dimensions ['season:JF']. expected either '...' to reduce over all dimensions or one or more of ('season', 'lat', 'lon').

**Reply by mathause** [reply]
You already reduce the time dimension in line 11 (`x = Data.groupby(seasons).mean()`) - i.e. no need to do it again but replace `mean` with `sum`.

However, you loose the the information on the years - do you rather need to use a `resample`-like functionality? Can you try `Data.resample(time="Q-DEC").mean()`? and see if this is more what you need. Of course this does not create your custom seasons - I am not sure how to do this.

**Comment by dcherian** [comment, marked as answer]
On latest Xarray (>2024.04.0) this works:
```python
import xarray as xr
from xarray.groupers import SeasonGrouper

ds = xr.tutorial.open_dataset("air_temperature")

ds.groupby(time=SeasonGrouper(["JF", "MAM", "JJAS", "OND"])).mean()
```

https://docs.xarray.dev/en/latest/user-guide/time-series.html#handling-seasons
